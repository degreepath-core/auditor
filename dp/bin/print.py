"""print

Given an input JSON student and the output of `dp --json` as argv, prints the output like dp would normally
"""

import json
import argparse

from dp.stringify import print_result
from dp import load_course


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("json_output", action='store')
    parser.add_argument("student_file", action='store')
    args = parser.parse_args()

    with open(args.json_output, 'r', encoding="utf-8") as infile:
        data = json.load(infile)

    with open(args.student_file, 'r', encoding="utf-8") as infile:
        student = json.load(infile)

    transcript = {
        c.clbid: c
        for c in (load_course(row, overrides=[]) for row in student["courses"])
    }

    summary = "\n".join(print_result(data, transcript=transcript))
    print(summary)


if __name__ == "__main__":
    main()
