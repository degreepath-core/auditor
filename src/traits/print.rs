use std::fmt;

pub type Result = std::result::Result<String, fmt::Error>;

pub trait Print {
	fn print(&self) -> Result;

	fn print_indented(&self, level: usize) -> Result {
		let printed = self.print()?;

		if printed.contains('\n') {
			Ok(indent_non_empty_lines(&printed, level))
		} else {
			Ok(printed.to_string())
		}
	}
}

fn indent_non_empty_lines(s: &str, times: usize) -> String {
	let lines: Vec<_> = s.lines().collect();

	if let Some((first, rest)) = lines.split_first() {
		let indent = "    ".repeat(times);

		// don't indent empty lines
		let rest = rest.iter().map(|l| {
			if l.trim() != "" {
				format!("{}{}", indent, l)
			} else {
				String::from("")
			}
		});

		// skip the first blank line
		let rest = rest.skip_while(|l| l.is_empty());

		let rest = rest.collect::<Vec<_>>().join("\n");

		format!("{}\n{}", first, rest)
	} else {
		s.to_string()
	}
}
