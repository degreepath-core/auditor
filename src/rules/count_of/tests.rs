use super::*;
use crate::traits::print::Print;
use insta::{assert_debug_snapshot_matches, assert_snapshot_matches, assert_yaml_snapshot_matches};

#[test]
fn serialize_count_of_any() {
	let data = Rule {
		count: Counter::Any,
		of: vec![],
	};

	assert_yaml_snapshot_matches!("serialize_count_of_any", data);
}

#[test]
fn deserialize_count_of_any() {
	let data = "---
count: any
of: []";

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_debug_snapshot_matches!("deserialize_count_of_any", actual);
}

#[test]
fn serialize_count_of_all() {
	let data = Rule {
		count: Counter::All,
		of: vec![],
	};

	assert_yaml_snapshot_matches!("serialize_count_of_all", data);
}

#[test]
fn deserialize_count_of_all() {
	let data = "---
count: all
of: []";

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_debug_snapshot_matches!("deserialize_count_of_all", actual);
}

#[test]
fn serialize_count_of_number() {
	let data = Rule {
		count: Counter::Number(6),
		of: vec![],
	};

	assert_yaml_snapshot_matches!("serialize_count_of_number", data);
}

#[test]
fn deserialize_count_of_number() {
	let data = "---
count: 6
of: []";

	let actual: Rule = serde_yaml::from_str(&data).unwrap();
	assert_debug_snapshot_matches!("deserialize_count_of_number", actual);
}

#[test]
fn pretty_print_inline() {
	let s = "{count: any, of: [CS 111]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take CS 111");

	let s = "{count: any, of: [{requirement: Core}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"complete the “Core” requirement");

	let s = "{count: any, of: [{both: [CS 111, CS 121]}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take both CS 111 and CS 121");

	let s = "{count: all, of: [CS 111, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take both CS 111 and CS 121");

	let s = "{count: any, of: [CS 111, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take either CS 111 or CS 121");

	let s = "{count: any, of: [CS 111, CS 121, CS 131]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take one course from among CS 111, CS 121, or CS 131");

	let s = "{count: all, of: [CS 111, CS 121, CS 131]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"take CS 111, CS 121, and CS 131");

	let s = "{count: all, of: [{requirement: A}, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"complete both the “A” and “B” requirements");

	let s = "{count: all, of: [{requirement: A}, {requirement: B}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"complete “A”, “B”, and “C”");

	let s = "{count: any, of: [{requirement: A}, {requirement: B}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"complete one requirement from among “A”, “B”, or “C”");

	let s = "{count: 2, of: [{requirement: A}, {requirement: B}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @"complete two requirements from among “A”, “B”, or “C”");
}

#[test]
fn all_two_course_one_requirement() {
	let expected = "take CS 111 and CS 121, and complete the “B” requirement";

	let s = "{count: all, of: [CS 111, CS 121, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: all, of: [CS 111, {requirement: B}, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: all, of: [{requirement: B}, CS 111, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn any_two_course_one_requirement() {
	let expected = "take CS 111 or CS 121, or complete the “B” requirement";

	let s = "{count: any, of: [CS 111, CS 121, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: any, of: [CS 111, {requirement: B}, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: any, of: [{requirement: B}, CS 111, CS 121]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn all_two_requirement_one_course() {
	let expected = "take CS 111, and complete both the “A” and “B” requirements";

	let s = "{count: all, of: [CS 111, {requirement: A}, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: all, of: [{requirement: A}, CS 111, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: all, of: [{requirement: A}, {requirement: B}, CS 111]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn any_two_requirement_one_course() {
	let expected = "take CS 111, or complete either the “A” or “B” requirements";

	let s = "{count: any, of: [CS 111, {requirement: A}, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: any, of: [{requirement: A}, CS 111, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: any, of: [{requirement: A}, {requirement: B}, CS 111]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn two_of_one_requirement_two_course() {
	let expected = "complete or take two requirements or courses from among CS 111, “A”, or “B”";

	let s = "{count: 2, of: [CS 111, {requirement: A}, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: 2, of: [{requirement: A}, CS 111, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: 2, of: [{requirement: A}, {requirement: B}, CS 111]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn two_of_two_requirement_one_course() {
	let expected = "complete or take two requirements or courses from among CS 111, “A”, or “B”";

	let s = "{count: 2, of: [CS 111, {requirement: A}, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: 2, of: [{requirement: A}, CS 111, {requirement: B}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);

	let s = "{count: 2, of: [{requirement: A}, {requirement: B}, CS 111]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_eq!(input.print().unwrap(), expected);
}

#[test]
fn both_either_requirement() {
	let s = "{count: all, of: [{both: [CS 111, CS 251]}, {either: [{requirement: A}, {requirement: B}]}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"do all of the following:

- take both CS 111 and CS 251
- complete either the “A” or “B” requirement
- complete the “C” requirement"###);
}

#[test]
fn pretty_print_block() {
	let s = "{count: any, of: [CS 111, CS 121, CS 124, CS 125]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take one of the following courses:

- CS 111
- CS 121
- CS 124
- CS 125"###);

	let s = "{count: 1, of: [CS 111, CS 121, CS 124, CS 125]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take one of the following courses:

- CS 111
- CS 121
- CS 124
- CS 125"###);

	let s = "{count: all, of: [CS 111, CS 121, CS 124, CS 125]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take all of the following courses:

- CS 111
- CS 121
- CS 124
- CS 125"###);

	let s = "{count: 4, of: [CS 111, CS 121, CS 124, CS 125]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take all of the following courses:

- CS 111
- CS 121
- CS 124
- CS 125"###);

	let s = "{count: 2, of: [CS 111, CS 121, CS 124, CS 125]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"take two of the following courses:

- CS 111
- CS 121
- CS 124
- CS 125"###);

	let s = "{count: any, of: [{requirement: A}, {requirement: B}, {requirement: C}, {requirement: D}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"complete one of the following requirements:

- “A”
- “B”
- “C”
- “D”"###);

	let s = "{count: any, of: [{requirement: A}, {requirement: B}, {requirement: C}, {requirement: D}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"complete one of the following requirements:

- “A”
- “B”
- “C”
- “D”"###);

	let s = "{count: 2, of: [{requirement: A}, {requirement: B}, {requirement: C}, {requirement: D}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"complete two from among the following requirements:

- “A”
- “B”
- “C”
- “D”"###);

	let s = "{count: 2, of: [{both: [CS 111, CS 251]}, {requirement: A}, {requirement: B}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"do two from among the following:

- take both CS 111 and CS 251
- complete the “A” requirement
- complete the “B” requirement
- complete the “C” requirement"###);

	let s = "{count: any, of: [{both: [CS 111, CS 251]}, {either: [{requirement: A}, {requirement: B}]}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"do one of the following:

- take both CS 111 and CS 251
- complete either the “A” or “B” requirement
- complete the “C” requirement"###);

	let s =
		"{count: 2, of: [{both: [CS 111, CS 251]}, {either: [{requirement: A}, {requirement: B}]}, {requirement: C}]}";
	let input: Rule = serde_yaml::from_str(&s).unwrap();
	assert_snapshot_matches!(input.print().unwrap(), @r###"do two from among the following:

- take both CS 111 and CS 251
- complete either the “A” or “B” requirement
- complete the “C” requirement"###);
}
