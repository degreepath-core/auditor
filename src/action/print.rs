use super::{Action, Command, Operator, Value};
use crate::traits::print;
use std::fmt::Write;

impl print::Print for Action {
	fn print(&self) -> print::Result {
		let mut output = String::new();

		match (&self.lhs, &self.op, &self.rhs) {
			(Value::String(s), Some(op), Some(val)) if *s == Command::Count => {
				write!(&mut output, "{} {}", op.print()?, val.print()?)?
			}
			(Value::String(s), Some(op), Some(val)) if *s == Command::Sum => {
				write!(&mut output, "{} {}", op.print()?, val.print()?)?
			}
			(Value::String(s), Some(Operator::GreaterThan), Some(val)) if *s == Command::Average => {
				write!(&mut output, "above {}", val.print()?)?
			}
			(Value::String(s), Some(Operator::GreaterThanEqualTo), Some(val)) if *s == Command::Average => {
				write!(&mut output, "at or above {}", val.print()?)?
			}
			_ => unimplemented!(
				"in Action's printer, the combo of `{:?}`, `{:?}`, and `{:?}`",
				&self.lhs,
				&self.op,
				&self.rhs
			),
		};

		Ok(output)
	}
}
