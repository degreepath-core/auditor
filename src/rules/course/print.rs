use super::*;
use crate::traits::print;
use crate::util;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "{}", self.course)?;

		if let Some(section) = &self.section {
			write!(&mut output, "{}", section)?;
		}

		let mut annotations = match (self.year, &self.semester) {
			(Some(year), None) => Some(util::expand_year(year, "dual")),
			(None, Some(semester)) => Some(semester.to_string()),
			(Some(year), Some(semester)) => Some(format!("{} {}", semester, util::expand_year(year, "short"))),
			(None, None) => None,
		};

		if let Some(lab) = self.lab {
			if lab {
				match &annotations {
					Some(ann) => {
						annotations = Some(format!("Lab; {}", ann));
					}
					None => {
						annotations = Some("Lab".to_string());
					}
				}
			}
		}

		match (&self.grade, &annotations) {
			(Some(grade), Some(ann)) => annotations = Some(format!("{}; {} or higher", ann, String::from(grade))),
			(Some(grade), None) => annotations = Some(format!("{} or higher", String::from(grade))),
			_ => (),
		}

		if let Some(annotations) = annotations {
			write!(&mut output, " ({})", annotations)?;
		}

		Ok(output)
	}
}
