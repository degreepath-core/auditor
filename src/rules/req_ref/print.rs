use super::*;
use crate::traits::print;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;
		let mut output = String::new();

		write!(&mut output, "“{}”", self.name)?;

		if self.optional {
			write!(&mut output, " (optional)")?;
		}

		Ok(output)
	}
}
