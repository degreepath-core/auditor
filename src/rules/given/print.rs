use super::{
	CourseRule, Given, GivenAreasWhatOptions, GivenAttendancesWhatOptions, GivenCoursesWhatOptions,
	GivenPerformancesWhatOptions, RepeatMode, Rule,
};
use crate::action;
use crate::filter::{AreaClause, AttendanceClause, CourseClause, PerformanceClause};
use crate::limit::Limiter;
use crate::traits::print::{self, Print};
use crate::util::Oxford;

impl print::Print for Rule {
	fn print(&self) -> print::Result {
		use std::fmt::Write;

		let mut output = String::new();

		let rule = match &self.given {
			Given::AllCourses { what, filter, limit } => self.print_given_all_courses(what, filter, limit)?,
			Given::TheseCourses {
				what,
				courses,
				repeats: mode,
				filter,
				limit,
			} => self.print_given_these_courses(courses, mode, what, filter, limit)?,
			Given::TheseRequirements {
				what,
				requirements,
				filter,
				limit,
			} => self.print_given_these_requirements(requirements, what, filter, limit)?,
			Given::NamedVariable {
				save,
				what,
				filter,
				limit,
			} => self.print_given_save(save, what, filter, limit)?,
			Given::Areas { what, filter } => self.print_given_areas(what, filter)?,
			Given::Performances { what, filter } => self.print_given_performances(what, filter)?,
			Given::Attendances { what, filter } => self.print_given_attendances(what, filter)?,
		};

		write!(&mut output, "{}", rule)?;

		Ok(output)
	}
}

impl Rule {
	fn print_given_all_courses(
		&self,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let action = self.action.print()?;
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};
		let limits = if let Some(limits) = &limit {
			let stringified = limits
				.iter()
				.map(|l| format!("at most {} {}", l.at_most, l.filter.print().unwrap()))
				.collect::<Vec<_>>()
				.oxford("and");
			format!(" {}", stringified)
		} else {
			"".to_owned()
		};

		match &what {
			What::Courses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "take {} {}{}{}", action, word, filter, limits)?;
			}
			What::DistinctCourses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "distinct courses" } else { "course" };

				write!(&mut output, "take {} {}{}", action, word, filter)?;
			}
			What::Credits => {
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(
					&mut output,
					"have enough courses{} to obtain {} {}",
					filter, action, word
				)?;
			}
			What::Departments => {
				let plur = self.action.should_pluralize();
				let word = if plur { "departments" } else { "department" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(&mut output, "have enough courses{} to span {} {}", filter, action, word)?;
			}
			What::Grades => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };
				let filter = if filter.is_empty() {
					"".to_owned()
				} else {
					format!(" taken{}", filter)
				};

				write!(
					&mut output,
					"maintain an average GPA {} from {}{}",
					action, word, filter
				)?;
			}
			What::Terms => {
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(
					&mut output,
					"have taken enough courses{} to span {} {}",
					filter, action, word
				)?;
			}
		}

		Ok(output)
	}

	fn print_given_areas(&self, what: &GivenAreasWhatOptions, filter: &Option<AreaClause>) -> print::Result {
		use std::fmt::Write;
		use GivenAreasWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Areas => {
				// TODO: find a better way to special-case "exactly one" major
				let action = self.action.print()?;
				let action = action.replace("exactly ", "");
				write!(&mut output, "declare {}{}", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_performances(
		&self,
		what: &GivenPerformancesWhatOptions,
		filter: &Option<PerformanceClause>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenPerformancesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Performances => {
				let action = self.action.print()?;
				write!(&mut output, "perform {} recitals{}", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_attendances(
		&self,
		what: &GivenAttendancesWhatOptions,
		filter: &Option<AttendanceClause>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenAttendancesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Attendances => {
				let action = self.action.print()?;
				write!(&mut output, "attend {}{} recitals", action, filter)?;
			}
		}

		Ok(output)
	}

	fn print_given_these_courses(
		&self,
		courses: &[CourseRule],
		mode: &RepeatMode,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => Some(format!(" {}", f.print()?)),
			None => None,
		};

		let courses: Vec<String> = courses.iter().map(|r| r.print().unwrap()).collect();

		match (mode, what) {
			(RepeatMode::First, What::Courses) | (RepeatMode::Last, What::Courses) => {
				match courses.len() {
					1 => {
						// TODO: expose last vs. first in output somehow?
						write!(&mut output, "take {}", courses.oxford("and"))?;
					}
					2 => match (&self.action.lhs, &self.action.op, &self.action.rhs) {
						(
							action::Command::Count,
							Some(action::Operator::GreaterThanEqualTo),
							Some(action::Value::Integer(n)),
						) => match n {
							1 => {
								write!(&mut output, "take either {} or {}", courses[0], courses[1])?;
							}
							2 => {
								write!(&mut output, "take both {} and {}", courses[0], courses[1])?;
							}
							_ => panic!("should not require <1 or >len of the number of courses given"),
						},
						_ => unimplemented!("most actions on two-up given:these-courses rules"),
					},
					3...5 => {
						// TODO: expose last vs. first in output somehow?
						let plur = self.action.should_pluralize();
						let word = if plur { "courses" } else { "course" };
						write!(
							&mut output,
							"take {} {} from among {}",
							self.action.print()?,
							word,
							courses.oxford("and")
						)?;
					}
					_ => {
						// TODO: expose last vs. first in output somehow?
						let plur = self.action.should_pluralize();
						let word = if plur { "courses" } else { "course" };

						let as_list: Vec<_> = courses.iter().map(|l| format!("- {}", l)).collect();

						write!(
							&mut output,
							"take {} {} from among the following:\n\n{}",
							self.action.print()?,
							word,
							as_list.join("\n")
						)?;
					}
				}
			}
			(RepeatMode::All, What::Courses) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "times" } else { "time" };

				match (&self.action.lhs, &self.action.op, &self.action.rhs) {
					(
						action::Command::Count,
						Some(action::Operator::GreaterThanEqualTo),
						Some(action::Value::Integer(1)),
					) => match courses.len() {
						1...5 => {
							write!(
								&mut output,
								"take {} {} {}",
								courses.oxford("or"),
								self.action.print()?,
								word
							)?;
						}
						_ => {
							let as_list: Vec<_> = courses.iter().map(|l| format!("- {}", l)).collect();

							write!(
								&mut output,
								"take {} of the following courses:\n\n{}",
								self.action.print()?,
								as_list.join("\n")
							)?;
						}
					},
					_ => match courses.len() {
						1 => {
							write!(
								&mut output,
								"take {} {} {}",
								courses.oxford("and"),
								self.action.print()?,
								word
							)?;
						}
						_ => {
							write!(
								&mut output,
								"take a combination of {} {} {}",
								courses.oxford("and"),
								self.action.print()?,
								word
							)?;
						}
					},
				}
			}
			(RepeatMode::All, What::Credits) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };

				write!(
					&mut output,
					"take {} enough times to yield {} {}",
					courses.oxford("and"),
					self.action.print()?,
					word
				)?;
			}
			(RepeatMode::All, What::Terms) => {
				// TODO: special-case "once" and "twice"
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(
					&mut output,
					"take {} enough times to span {} {}",
					courses.oxford("and"),
					self.action.print()?,
					word
				)?;
			}
			_ => unimplemented!("certain modes of given:these-courses"),
		}

		if let Some(f) = filter {
			write!(&mut output, "{}", f)?;
		}

		Ok(output)
	}

	fn print_given_these_requirements(
		&self,
		requirements: &[String],
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();

		let requirements: Vec<String> = requirements.iter().map(|r| format!("“{}”", r)).collect();

		writeln!(&mut output, "have the following be true:\n")?;
		let mut index = 0;

		index += 1;
		match requirements.len() {
			0 => panic!("no requirements given!"),
			1 | 2 | 3 => {
				let singular = requirements.len() == 1;
				let word = if singular { "requirement" } else { "requirements" };
				writeln!(
					&mut output,
					"{index}. given the results of the {list} {word},",
					index = index,
					list = requirements.oxford("and"),
					word = word
				)?;
			}
			_ => {
				writeln!(
					&mut output,
					"{index}. given the results of the following requirements",
					index = index
				)?;
				for req in requirements {
					writeln!(&mut output, "    - {}", req)?;
				}
			}
		};

		match &filter {
			Some(f) => {
				index += 1;
				writeln!(
					&mut output,
					"{index}. restricted to only courses taken {filter},",
					index = index,
					filter = f.print()?
				)?;
			}
			None => (),
		};

		index += 1;
		let pluralize = self.action.should_pluralize();

		match &what {
			What::Courses => {
				let word = if pluralize { "courses" } else { "course" };

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::DistinctCourses => {
				let word = if pluralize {
					"distinct courses"
				} else {
					"distinct course"
				};

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Credits => {
				let word = if pluralize { "credits" } else { "credit" };

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Departments => {
				let word = if pluralize {
					"distinct departments"
				} else {
					"department"
				};

				writeln!(
					&mut output,
					"{index}. there must be {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
			What::Grades => {
				writeln!(
					&mut output,
					"{index}. there must be an average GPA {action}",
					index = index,
					action = self.action.print()?,
				)?;
			}
			What::Terms => {
				let word = if pluralize { "terms" } else { "term" };

				writeln!(
					&mut output,
					"{index}. there must be courses in {action} {word}",
					index = index,
					action = self.action.print()?,
					word = word,
				)?;
			}
		};

		Ok(output)
	}

	fn print_given_save(
		&self,
		save: &str,
		what: &GivenCoursesWhatOptions,
		filter: &Option<CourseClause>,
		_limit: &Option<Vec<Limiter>>,
	) -> print::Result {
		use std::fmt::Write;
		use GivenCoursesWhatOptions as What;

		let mut output = String::new();
		let filter = match &filter {
			Some(f) => format!(" taken {}", f.print()?),
			None => "".to_string(),
		};

		match &what {
			What::Courses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(&mut output, "there must be {} {}{}", self.action.print()?, word, filter)?;
				// write!(&mut output, " in the subset “{}”", save)?;
			}
			What::DistinctCourses => {
				let plur = self.action.should_pluralize();
				let word = if plur { "distinct courses" } else { "course" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(&mut output, "there must be {} {}{}", self.action.print()?, word, filter)?;
				// write!(&mut output, " in the subset “{}”", save)?;
			}
			What::Credits => {
				let plur = self.action.should_pluralize();
				let word = if plur { "credits" } else { "credit" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to obtain {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
			What::Departments => {
				let plur = self.action.should_pluralize();
				let word = if plur { "departments" } else { "department" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to span {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
			What::Grades => {
				let plur = self.action.should_pluralize();
				let word = if plur { "courses" } else { "course" };

				write!(&mut output, "courses from the subset “{}” ", save)?;
				write!(
					&mut output,
					"must maintain an average GPA {} from {}{}",
					self.action.print()?,
					word,
					filter
				)?;
			}
			What::Terms => {
				let plur = self.action.should_pluralize();
				let word = if plur { "terms" } else { "term" };

				write!(&mut output, "in the subset “{}”, ", save)?;
				write!(
					&mut output,
					"there must be enough courses{} to span {} {}",
					filter,
					self.action.print()?,
					word
				)?;
			}
		}

		Ok(output)
	}
}
