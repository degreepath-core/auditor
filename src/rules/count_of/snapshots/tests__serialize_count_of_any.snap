---
created: "2019-02-27T18:19:43.102148Z"
creator: insta@0.6.3
source: src/rules/count_of/tests.rs
expression: data
---
count: any
of: []
